core/drivers/stm32_cpu_opp.c

```diff
@@ -0,0 +1,354 @@
+		mutex_unlock(&cpu_opp_mu);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (level < current_level)
+		res = set_clock_then_voltage(opp);
+	else
+		res = set_voltage_then_clock(opp);
+
+	if (res)
+		EMSG("Failed to set OPP to %ukHz", level);
+	else
+		cpu_opp.current_opp = opp;
+
+	mutex_unlock(&cpu_opp_mu);
+
+	return res;
+}
+
+TEE_Result stm32_cpu_opp_read_level(unsigned int *level)
+{
+	*level = stm32_cpu_opp_level(cpu_opp.current_opp);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
+{
+	const fdt64_t *cuint64 = NULL;
+	const fdt32_t *cuint32 = NULL;
+	uint64_t freq_hz = 0;
+	uint64_t freq_khz = 0;
+	uint64_t freq_khz_opp_def = 0;
+	uint32_t volt_uv = 0;
+	unsigned long clk_cpu = 0;
+	unsigned int i = 0;
+	int subnode = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	cpu_opp.opp_count = CFG_STM32MP_OPP_COUNT;
+
+	cpu_opp.dvfs = calloc(cpu_opp.opp_count, sizeof(*cpu_opp.dvfs));
+	if (!cpu_opp.dvfs)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		cuint64 = fdt_getprop(fdt, subnode, "opp-hz", NULL);
+		if (!cuint64) {
+			EMSG("Missing opp-hz in node %s",
+			     fdt_get_name(fdt, subnode, NULL));
+			res = TEE_ERROR_GENERIC;
+			goto err;
+		}
+
+		freq_hz = fdt64_to_cpu(*cuint64);
+		freq_khz = freq_hz / ULL(1000);
+		if (freq_khz > (uint64_t)UINT32_MAX) {
+			EMSG("Invalid opp-hz %"PRIu64" in node %s",
+			     freq_khz, fdt_get_name(fdt, subnode, NULL));
+			res = TEE_ERROR_GENERIC;
+			goto err;
+		}
+
+		cuint32 = fdt_getprop(fdt, subnode, "opp-microvolt", NULL);
+		if (!cuint32) {
+			EMSG("Missing opp-microvolt in node %s",
+			     fdt_get_name(fdt, subnode, NULL));
+			res = TEE_ERROR_GENERIC;
+			goto err;
+		}
+
+		volt_uv = fdt32_to_cpu(*cuint32);
+
+		if (i == cpu_opp.opp_count) {
+			EMSG("Too many OPP defined in node %s",
+			     fdt_get_name(fdt, node, NULL));
+			res = TEE_ERROR_GENERIC;
+			goto err;
+		}
+
+		cpu_opp.dvfs[i].freq_khz = freq_khz;
+		cpu_opp.dvfs[i].volt_uv = volt_uv;
+
+		DMSG("Found OPP %u (%"PRIu64"kHz/%"PRIu32"uV) from DT",
+		     i, freq_khz, volt_uv);
+
+		/* Select the max "st,opp-default" node as current OPP */
+		if (fdt_getprop(fdt, subnode, "st,opp-default", NULL) &&
+		    freq_khz > freq_khz_opp_def) {
+			cpu_opp.current_opp = i;
+			freq_khz_opp_def = freq_khz;
+		}
+
+		i++;
+	}
+
+	/* At least one OPP node shall have a "st,opp-default" property */
+	if (freq_khz_opp_def == 0) {
+		EMSG("No default OPP found");
+		res = TEE_ERROR_GENERIC;
+		goto err;
+	}
+
+	/* Apply the current OPP */
+	DMSG("Set OPP to %"PRIu64"kHz", freq_khz_opp_def);
+	clk_cpu = clk_get_rate(cpu_opp.clock);
+	assert(clk_cpu);
+	if (freq_khz_opp_def * ULL(1000) > clk_cpu)
+		res = set_voltage_then_clock(cpu_opp.current_opp);
+	else
+		res = set_clock_then_voltage(cpu_opp.current_opp);
+
+	if (res) {
+		EMSG("Failed to set default OPP %u", cpu_opp.current_opp);
+		goto err;
+	}
+
+	return TEE_SUCCESS;
+
+err:
```

Comment:
Maybe reset `cpu_opp.opp_count` as a hit for OPPs not being defined yet. If so, consider also the `calloc()` error case at function entry.